hkmc2/shared/src/main/scala/hkmc2/syntax/Tree.scala

```diff
@@ -239,35 +244,97 @@ enum Tree extends AutoLocated:
     
     case _ => this
   
-  /** 
-   * Parameter `inUsing` means the param list is modified by `using`.
-   * In the first result, `S(true)` means eager spread, `S(false)` means lazy spread, and `N` means no spread.
+  /**
+   * Parse a tree as a parameter.
+   * @param inUsing whether the parameter is in a `using` parameter list
+   * @param inDataClass whether the parameter is in a data class
    */
-  def asParam(inUsing: Bool): Diagnostic \/ (Opt[Bool], Ident, Opt[Tree]) = this match
-    case und: Under => R(N, new Ident("_").withLocOf(und), N)
-    // * In `using` clauses, identifiers and type applications are
-    // * understood as type names for unnamed contextual parameters:
-    case ty: Ident if inUsing => R(N, Ident(""), S(ty))
-    case ty @ TyApp(_, _) if inUsing => R(N, Ident(""), S(ty))
-    case id: Ident => R(N, id, N)
-    case Spread(Keyword.`..`, _, S(id: Ident)) => R(S(false), id, N)
-    case Spread(Keyword.`...`, _, S(id: Ident)) => R(S(true), id, N)
-    case Spread(Keyword.`..`, _, S(und: Under)) => R(S(false), new Ident("_").withLocOf(und), N)
-    case Spread(Keyword.`...`, _, S(und: Under)) => R(S(true), new Ident("_").withLocOf(und), N)
-    case Spread(Keyword.`...`, kwLoc, N) => R(S(true), new Ident("_").withLoc(kwLoc), N)
-    case Spread(Keyword.`..`, kwLoc, N) => R(S(false), new Ident("_").withLoc(kwLoc), N)
-    case InfixApp(lhs: Ident, Keyword.`:`, rhs) => R(N, lhs, S(rhs))
-    case TermDef(ImmutVal | MutVal, inner, _) => inner.asParam(inUsing)
-    case TermDef(Ins, inner, N) => inner.asParam(inUsing)
-    case _ => L:
-      ErrorReport:
-        msg"Expected a valid parameter, found ${this.describe}" -> this.toLoc :: Nil
-  
-  def isModuleModifier: Bool = this match
-    case td @ Tree.TypeDef(Mod, _, rhs) => rhs.isEmpty && td.extension.isEmpty && td.withPart.isEmpty
-    case _ => false
+  def asParam(inUsing: Bool, inDataClass: Bool): Diagnostic \/ TreeParam =
+    @tailrec
+    def go(t: Tree, flags: FldFlags, modifiers: Set[DeclKind]): Diagnostic \/ TreeParam = t match
+      // * Base Cases.
+      
+      // fun f(_)
+      case und: Under => 
+        R(TreeParam(flags, new Ident("_").withLocOf(und), N, N, modifiers))
+
+      // In `using` clauses, identifiers and type applications are
+      // understood as type names for unnamed contextual parameters:
```

Comment:
I think that instead of special-casing identifiers and type applications, you should special case `InfixApp(_, Keyword.`:`, _)`. You shouldn't make such strict assumptions about the shape of a type tree. We can very well have type trees of other shapes, like `OpApp`.